source/Mlos.Proto/OptimizerService.proto

```diff
@@ -67,26 +68,21 @@ message RegisterObservationsRequest
     Observations Observations = 2;
 };
 
-// A list of observations. TODO: make more sensible the difference between Observation and Observations. 
-// Specifically - they can probably be the same thing, but changing the name of Observation to Observations would
-// break Greg's PR. 
-// 
-message Observations
-{
-    Features Features = 1;
-    ObjectiveValues ObjectiveValues = 2;
-};
 
 // A representation of an optimization problem.
 //
 message OptimizationProblem
 {
-    Hypergrid ParameterSpace = 1;
-    Hypergrid ContextSpace = 2;
-    Hypergrid ObjectiveSpace = 3;
-    repeated Objective Objectives = 4;
+    SimpleHypergrid ParameterSpace = 1;
+    oneof ContextOptions {

```

Comment:
Why is this called ContextOptions instead of just context? Also I think gRPC allowed for an `optional` field which should make it nullable on the C# side.

Reply:
My understanding is that gRPC 3.0 makes all fields optional

The problem is that if no value is passed on the python side, then the gRPC instance will create an empty hypergrid as the parameter for the context for the OptimizerServiceDecoder(). Now you don't know whether or not the parameter was null to begin with or just an empty hypergrid. (similar but not identical and may not always be. Bad assumption to make). 

My solution was to include an option for an empty value via a oneof{} structure. 

This solution has the benefit of working and being resilient at the price of not being as elegant as a null context field. 

Do you have other suggestions or thoughts?

Reply:
Updated GrpcInterface to be a bit cleaner (No Empty value)
188f02e

Reply:
Thoughts?

Reply:
https://github.com/protocolbuffers/protobuf/releases/tag/v3.15.0
Optional fields for proto3 are enabled by default, and no longer require
the --experimental_allow_proto3_optional flag.
why not simply use optional Hypergrid ContextSpace?

Reply:
ContextSpace field grpc
af8326e

https://github.com/protocolbuffers/protobuf/blob/master/docs/field_presence.md
Sometimes SO isn't always right.